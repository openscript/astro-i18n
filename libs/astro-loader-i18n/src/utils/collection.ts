export function getAllUniqueKeys(obj: Record<string, unknown>, keys = new Set<string>()): Set<string> {
  Object.entries(obj).forEach(([key, value]) => {
    keys.add(key);
    if (value && typeof value === "object" && !Array.isArray(value)) {
      getAllUniqueKeys(value as Record<string, unknown>, keys);
    }
  });
  return keys;
}

function recursivePruneLocales(obj: Record<string, unknown>, locales: string[], locale: string) {
  const result: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(obj)) {
    if (Array.isArray(value)) {
      result[key] = value.map((item) => {
        if (item && typeof item === "object" && !Array.isArray(item)) {
          return recursivePruneLocales(item as Record<string, unknown>, locales, locale);
        }
        return item;
      });
    } else if (value && typeof value === "object") {
      const valueAsRecord = value as Record<string, unknown>;
      const hasLocales = Object.keys(valueAsRecord).some((k) => locales.includes(k));
      let prunedValue: unknown | undefined = undefined;

      if (hasLocales) {
        prunedValue = valueAsRecord[locale] ?? undefined;
      } else {
        prunedValue = recursivePruneLocales(valueAsRecord, locales, locale);
      }

      result[key] = prunedValue;
    } else {
      result[key] = value;
    }
  }

  return result;
}

export function pruneLocales(obj: Record<string, unknown>, locales: string[], locale: string) {
  if (Object.keys(obj).find((key) => locales.includes(key))) throw new Error("Top-level locales are not allowed");
  if (Object.keys(obj).length === 0) return obj;

  return recursivePruneLocales(obj, locales, locale);
}
