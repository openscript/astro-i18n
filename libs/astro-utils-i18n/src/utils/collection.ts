export function getAllUniqueKeys(obj: Record<string, unknown>, keys = new Set<string>(), ignore?: boolean): Set<string> {
  Object.entries(obj).forEach(([key, value]) => {
    if (!ignore) keys.add(key);
    if (value && typeof value === "object") {
      if (Array.isArray(value)) {
        value.forEach((item) => {
          if (item && typeof item === "object") {
            // Recurse into objects and arrays within arrays
            if (Array.isArray(item)) {
              // Wrap array in an object to reuse the function and ignore its key
              getAllUniqueKeys({ array: item }, keys, true);
            } else {
              getAllUniqueKeys(item as Record<string, unknown>, keys);
            }
          }
        });
      } else {
        getAllUniqueKeys(value as Record<string, unknown>, keys);
      }
    }
  });
  return keys;
}

function recursivePruneLocales(obj: Record<string, unknown>, locales: string[], locale: string) {
  const result: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(obj)) {
    if (Array.isArray(value)) {
      result[key] = value.map((item) => {
        if (item && typeof item === "object" && !Array.isArray(item)) {
          return recursivePruneLocales(item as Record<string, unknown>, locales, locale);
        }
        return item;
      });
    } else if (value && typeof value === "object") {
      const valueAsRecord = value as Record<string, unknown>;
      const hasLocales = Object.keys(valueAsRecord).some((k) => locales.includes(k));
      let prunedValue: unknown | undefined = undefined;

      if (hasLocales) {
        prunedValue = valueAsRecord[locale] ?? undefined;
      } else {
        prunedValue = recursivePruneLocales(valueAsRecord, locales, locale);
      }

      result[key] = prunedValue;
    } else {
      result[key] = value;
    }
  }

  return result;
}

export function pruneLocales(obj: Record<string, unknown>, locales: string[], locale: string) {
  if (Object.keys(obj).find((key) => locales.includes(key))) throw new Error("Top-level locales are not allowed");
  if (Object.keys(obj).length === 0) return obj;

  return recursivePruneLocales(obj, locales, locale);
}
