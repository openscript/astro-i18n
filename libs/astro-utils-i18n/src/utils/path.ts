const TRIM_SLASHES_PATTERN = /^\/|\/$/g;
const TRIM_RELATIVE_PATTERN = /^\.{0,2}\/|\/$/g;
const DOUBLE_POINTS_PATTERN = /\.{2}/g;
const DOUBLE_SLASH_PATTERN = /\/{2}/g;
const DIRNAME_PATTERN = /\/[^/]+$/g;
const INDEX_COMMON_TRANSLATION_ID = "index";
const createLocalePattern = (pathLocale: string) => new RegExp(`(?<=[./])${pathLocale}(?=[./])`, "i");

export function joinPath(...paths: Array<string | number | undefined>) {
  return paths.filter(Boolean).join("/");
}

/**
 * Resolves and joins multiple path segments into a single normalized path.
 *
 * This function trims leading and trailing slashes from string path segments
 * and ensures the resulting path starts with a single forward slash (`/`).
 * Non-string segments (e.g., numbers or `undefined`) are included as-is.
 *
 * @param paths - An array of path segments which can be strings, numbers, or `undefined`.
 * @returns A normalized path string starting with a forward slash.
 */
export function resolvePath(...paths: Array<string | number | undefined>) {
  const trimmedPaths = paths.map((path) => (typeof path === "string" ? path.replace(TRIM_SLASHES_PATTERN, "") : path));
  return `/${joinPath(...trimmedPaths)}`;
}

export const trimSlashes = (path: string) => {
  return path === "/" ? path : path.replace(TRIM_SLASHES_PATTERN, "");
};

export const trimRelativePath = (path: string) => {
  return path === "/" ? path : path.replace(TRIM_RELATIVE_PATTERN, "");
};

export const parseLocale = (path: string, locales: string[], defaultLocale: string) => {
  const localePattern = createLocalePattern(`(${locales.join("|")})`);
  const locale = path.match(localePattern)?.[0];

  return locale ? locale : defaultLocale;
};

export const createTranslationId = (path: string, locale?: string) => {
  if (locale) path = path.replace(createLocalePattern(locale), "");
  path = path.replace(DOUBLE_POINTS_PATTERN, ".");
  path = path.replace(DOUBLE_SLASH_PATTERN, "/");
  path = trimSlashes(path);

  return path === "" || path === "/" ? INDEX_COMMON_TRANSLATION_ID : path;
};

export const createContentPath = (path: string, base?: string | URL, locale?: string) => {
  let basePath = base ? (base instanceof URL ? base.pathname : base) : undefined;
  if (basePath) {
    basePath = trimRelativePath(basePath);
    const basePathIndex = path.indexOf(basePath);
    path = basePathIndex !== -1 ? path.slice(basePathIndex + basePath.length) : path;
  }
  path = createTranslationId(path, locale);
  if (path.includes(INDEX_COMMON_TRANSLATION_ID)) {
    path = path.replace(DIRNAME_PATTERN, "");
  }
  path = path.includes("/") ? path.replace(DIRNAME_PATTERN, "") : "";

  return path;
};
