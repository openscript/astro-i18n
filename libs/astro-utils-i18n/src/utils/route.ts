import { resolvePath } from "./path";

type Segments = Record<string, string | undefined>;
export type SegmentTranslations = Record<string, Segments>;
type RouteSegment = {
  spread: boolean;
  param: boolean;
  value: string;
};
type RoutePattern = RouteSegment[];

/**
 * Parses a route pattern into an array of route segments.
 *
 * @param routePattern is a string that templates the a route (e.g. /[...locale]/blog/[slug])
 */
export function parseRoutePattern(routePattern: string): RoutePattern {
  const segments = routePattern
    .split("/")
    .filter((segment) => segment !== "")
    .map((segment) => {
      if (segment.startsWith("[") && segment.endsWith("]")) {
        return {
          spread: segment.startsWith("[..."),
          param: true,
          value: segment.replace("[", "").replace("...", "").replace("]", ""),
        };
      }

      return {
        spread: false,
        param: false,
        value: segment,
      };
    });

  if (segments.length === 0) {
    return [{ spread: false, param: false, value: "/" }];
  }

  return segments;
}

/**
 * Constructs a full path by resolving a base path with a route pattern and corresponding segment values.
 *
 * @param routePattern - An array of route segments that define the structure of the route.
 *                        Each segment can either be a static value or a parameter.
 * @param segmentValues - An object mapping parameter names to their corresponding values.
 *                        These values are used to replace parameterized segments in the route pattern.
 * @param basePath - The base path to resolve the constructed route against.
 *
 * @returns The resolved path as a string.
 *
 * @throws {Error} If a required segment value for a parameterized route segment is missing
 *                 and the segment is not marked as a spread segment.
 */
export function buildPath(routePattern: RoutePattern, segmentValues: Segments, basePath: string) {
  return resolvePath(
    basePath,
    ...routePattern.map((segment) => {
      if (segment.param) {
        if (!segmentValues[segment.value] && !segment.spread) {
          throw new Error(`No segment value found for route segment "${segment.value}". Did you forget to provide it?`);
        }
        if (segmentValues[segment.value]?.includes("/") && !segment.spread) {
          throw new Error(
            `The segment value "${segmentValues[segment.value]}" for route segment "${segment.value}" contains a slash. Did you forget to add "..." to the route pattern?`
          );
        }

        return segmentValues[segment.value];
      }
      return `${segment.value}`;
    })
  );
}
