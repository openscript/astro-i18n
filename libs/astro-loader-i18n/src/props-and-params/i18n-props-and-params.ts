import limax from "limax";
import { checkI18nLoaderCollection, i18nLoaderSchema } from "../schemas/i18n-loader-schema";
import { buildPath, parseRoutePattern, SegmentTranslations } from "../utils/route";
import { z } from "astro/zod";
import { CollectionEntry, CollectionKey } from "astro:content";
import { I18nCollection } from "../collections/create-i18n-collection";
import { joinPath } from "../utils/path";

type Config = {
  defaultLocale: string;
  routePattern: string;
  segmentTranslations: SegmentTranslations;
  localeParamName?: string;
  slugParamName?: string;
  titleDataKey?: string;
  prefixDefaultLocale?: boolean;
};

const defaultConfig = {
  localeParamName: "locale",
  slugParamName: "slug",
  titleDataKey: "title",
  prefixDefaultLocale: false,
} as const;

function getSegmentTranslations(
  data: z.infer<typeof i18nLoaderSchema> & Record<string, string | unknown>,
  c: Omit<Required<Config>, "routePattern">
) {
  if (!c.segmentTranslations[data.locale]) throw new Error(`No slugs found for locale ${data.locale}`);

  const currentLocale = !c.prefixDefaultLocale && data.locale === c.defaultLocale ? undefined : data.locale;
  const segmentValues = { [c.localeParamName]: currentLocale, ...c.segmentTranslations[data.locale] };

  const slugValue = c.titleDataKey ? data[c.titleDataKey] : undefined;
  if (slugValue && typeof slugValue === "string") {
    segmentValues[c.slugParamName] = joinPath(data.contentPath, limax(slugValue));
  }
  return segmentValues;
}

/**
 * Processes a collection of entries and generates i18n-related properties and parameters.
 *
 * @template C - The type of the collection entries.
 * @param collection - The collection of entries or an i18n collection.
 * @param config - The configuration object for i18n processing.
 * @returns An array of objects containing `params` and `props` for each entry.
 * @throws {Error} If route segment translations or slug parameters are invalid.
 */
export function i18nPropsAndParams<C extends CollectionEntry<CollectionKey>[]>(collection: C | I18nCollection, config: Config) {
  checkI18nLoaderCollection(collection);
  const { routePattern, ...c } = { ...defaultConfig, ...config };
  const route = parseRoutePattern(routePattern);

  route.forEach((segment, index) => {
    if (
      segment.param &&
      segment.value !== c.localeParamName &&
      index !== route.length - 1 &&
      !Object.values(c.segmentTranslations).every((translation) => translation[segment.value])
    ) {
      throw new Error(`No slugs found for route segment ${segment.value}`);
    }
    if (segment.value === c.slugParamName && !segment.spread && collection.some((entry) => entry.data.contentPath !== "")) {
      const example = collection.find((entry) => entry.data.contentPath !== "");
      throw new Error(
        `The slug param "${segment.value}" requires to be spread, because some entries (e.g ${example?.data.translationId}:${example?.data.contentPath}) have a contentPath, that leads to a slug with a path. Add "..." to the slug param in your route pattern ${routePattern} via changing the corresponding file or folder name.`
      );
    }
  });

  return collection.map((entry) => {
    const { translationId } = entry.data;
    const entryTranslations = collection.filter((e) => e.data.translationId === translationId);
    const translations = entryTranslations.reduce(
      (previous, current) => {
        return {
          ...previous,
          [current.data.locale]: buildPath(route, getSegmentTranslations(current.data, c), current.data.basePath),
        };
      },
      {} as Record<string, string>
    );

    const params = getSegmentTranslations(entry.data, c);
    const translatedPath = buildPath(route, params, entry.data.basePath);

    return {
      params,
      props: {
        ...entry,
        translations,
        translatedPath,
      } as C[number] & I18nCollection[number] & { translations: Record<string, string>; translatedPath: string },
    };
  });
}
