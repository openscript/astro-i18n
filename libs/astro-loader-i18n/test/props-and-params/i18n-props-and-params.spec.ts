import { describe, expect, it } from "vitest";
import { i18nPropsAndParams } from "../../src/props-and-params/i18n-props-and-params";
import { SegmentTranslations } from "../../src/utils/route";

const COLLECTION_FIXTURE = [
  {
    data: { locale: "de-CH", translationId: "magic.mdx", contentPath: "test", basePath: "/", title: "Verückte Umlaute!" },
  },
  {
    data: { locale: "zh-CN", translationId: "magic.mdx", contentPath: "test", basePath: "/", title: "神奇的标题" },
  },
];

describe("i18nPropsAndParams", () => {
  it("should generate paths for a valid collection", () => {
    const routePattern = "[...locale]/[blog]/posts/[...slug]";
    const segmentTranslations: SegmentTranslations = { "de-CH": { blog: "logbuch" }, "zh-CN": { blog: "blog" } };
    const result = i18nPropsAndParams(COLLECTION_FIXTURE, { routePattern, segmentTranslations, defaultLocale: "de-CH" });
    expect(result).toMatchSnapshot();
  });
  it("should generate full paths if prefixDefaultLocale is true", () => {
    const routePattern = "[...locale]/[blog]/posts/[...slug]";
    const segmentTranslations: SegmentTranslations = { "de-CH": { blog: "logbuch" }, "zh-CN": { blog: "blog" } };
    const result = i18nPropsAndParams(COLLECTION_FIXTURE, {
      routePattern,
      segmentTranslations,
      defaultLocale: "de-CH",
      prefixDefaultLocale: true,
    });
    expect(result).toMatchSnapshot();
  });
  it("should should throw an error if the slug needs to be a spread param, but isn't", () => {
    const routePattern = "[...locale]/[blog]/posts/[slug]";
    const segmentTranslations: SegmentTranslations = { "de-CH": { blog: "logbuch" }, "zh-CN": { blog: "blog" } };
    expect(() =>
      i18nPropsAndParams(COLLECTION_FIXTURE, { routePattern, segmentTranslations, defaultLocale: "de-CH" })
    ).toThrowErrorMatchingSnapshot();
  });
  it("should throw an error if not all params can be filled", () => {
    const routePattern = "[...locale]/[blog]/[unknown]/posts/[slug]";
    const segmentTranslations: SegmentTranslations = { "de-CH": { blog: "logbuch" }, "zh-CN": { blog: "blog" } };
    expect(() =>
      i18nPropsAndParams(COLLECTION_FIXTURE, { routePattern, segmentTranslations, defaultLocale: "de-CH" })
    ).toThrowErrorMatchingSnapshot();
  });
});
