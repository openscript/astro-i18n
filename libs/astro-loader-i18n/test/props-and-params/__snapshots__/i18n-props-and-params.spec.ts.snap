// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`i18nPropsAndParams > should generate full paths if prefixDefaultLocale is true 1`] = `
[
  {
    "params": {
      "blog": "logbuch",
      "locale": "de-CH",
      "slug": "test/veruckte-umlaute",
    },
    "props": {
      "data": {
        "basePath": "/",
        "contentPath": "test",
        "locale": "de-CH",
        "title": "Verückte Umlaute!",
        "translationId": "magic.mdx",
      },
      "translatedPath": "/de-CH/logbuch/posts/test/veruckte-umlaute",
      "translations": {
        "de-CH": "/de-CH/logbuch/posts/test/veruckte-umlaute",
        "zh-CN": "/zh-CN/blog/posts/test/shen2-qi2-de-biao1-ti2",
      },
    },
  },
  {
    "params": {
      "blog": "blog",
      "locale": "zh-CN",
      "slug": "test/shen2-qi2-de-biao1-ti2",
    },
    "props": {
      "data": {
        "basePath": "/",
        "contentPath": "test",
        "locale": "zh-CN",
        "title": "神奇的标题",
        "translationId": "magic.mdx",
      },
      "translatedPath": "/zh-CN/blog/posts/test/shen2-qi2-de-biao1-ti2",
      "translations": {
        "de-CH": "/de-CH/logbuch/posts/test/veruckte-umlaute",
        "zh-CN": "/zh-CN/blog/posts/test/shen2-qi2-de-biao1-ti2",
      },
    },
  },
]
`;

exports[`i18nPropsAndParams > should generate paths for a valid collection 1`] = `
[
  {
    "params": {
      "blog": "logbuch",
      "locale": undefined,
      "slug": "test/veruckte-umlaute",
    },
    "props": {
      "data": {
        "basePath": "/",
        "contentPath": "test",
        "locale": "de-CH",
        "title": "Verückte Umlaute!",
        "translationId": "magic.mdx",
      },
      "translatedPath": "/logbuch/posts/test/veruckte-umlaute",
      "translations": {
        "de-CH": "/logbuch/posts/test/veruckte-umlaute",
        "zh-CN": "/zh-CN/blog/posts/test/shen2-qi2-de-biao1-ti2",
      },
    },
  },
  {
    "params": {
      "blog": "blog",
      "locale": "zh-CN",
      "slug": "test/shen2-qi2-de-biao1-ti2",
    },
    "props": {
      "data": {
        "basePath": "/",
        "contentPath": "test",
        "locale": "zh-CN",
        "title": "神奇的标题",
        "translationId": "magic.mdx",
      },
      "translatedPath": "/zh-CN/blog/posts/test/shen2-qi2-de-biao1-ti2",
      "translations": {
        "de-CH": "/logbuch/posts/test/veruckte-umlaute",
        "zh-CN": "/zh-CN/blog/posts/test/shen2-qi2-de-biao1-ti2",
      },
    },
  },
]
`;

exports[`i18nPropsAndParams > should should throw an error if the slug needs to be a spread param, but isn't 1`] = `[Error: The slug param "slug" requires to be spread, because some entries (e.g magic.mdx:test) have a contentPath, that leads to a slug with a path. Add "..." to the slug param in your route pattern [...locale]/[blog]/posts/[slug] via changing the corresponding file or folder name.]`;

exports[`i18nPropsAndParams > should throw an error if not all params can be filled 1`] = `[Error: No slugs found for route segment unknown]`;
